eth_utils/abi.py

```diff
@@ -50,6 +80,188 @@ def _abi_to_signature(abi: Dict[str, Any]) -> str:
+        raise ValidationError("Expected non-anonymous event to have 1 or more topics")
+    elif event_abi_to_log_topic(dict(event_abi)) != _log_entry_data_to_bytes(topics[0]):
+        # raise MismatchedABI("The event signature did not match the provided ABI")
+        raise ValidationError("The event signature did not match the provided ABI")
+    else:
+        return topics[1:]
+
+
+def get_all_event_abis(abi: ABI) -> Sequence[ABIEvent]:
+    """
+    Return interfaces for each event in the contract ABI.
+
+    :param abi: Contract ABI.
+    :param type: `ABI`
+    :return: List of ABIs for each event interface.
+    :rtype: `list[ABIEvent]`
+    """
+    return [ABIEvent(event) for event in _filter_by_type("event", abi)]
+
+
+def get_event_abi(
+    abi: ABI,
+    event_name: Optional[str] = None,
+    argument_names: Optional[Sequence[str]] = None,
+) -> ABIEvent:
+    """
+    Find the event interface with the given name and arguments.
+
+    :param abi: Contract ABI.
+    :param type: `ABI`
+    :param event_name: Find an event abi with matching event name.
+    :param type: `str`
+    :param argument_names: Find an event abi with matching arguments.
+    :param type: `list[str]`
+    :return: ABI for the event interface.
+    :rtype: `ABIEvent`
+    """
+    filters = [
+        functools.partial(_filter_by_type, "event"),
+    ]
+
+    if event_name is not None:
+        filters.append(functools.partial(_filter_by_name, event_name))
+
+    if argument_names is not None:
+        filters.append(functools.partial(_filter_by_argument_name, argument_names))
+
+    event_abi_candidates = pipe(abi, *filters)
+
+    if len(event_abi_candidates) == 1:
+        return event_abi_candidates[0]
+    elif len(event_abi_candidates) == 0:
+        raise ValueError("No matching events found")
+    else:
+        raise ValueError("Multiple events found")
+
+
+def get_abi_input_names(abi_element: ABIElement) -> List[str]:
+    """
+    Return names for each input from the function or event ABI.
+
+    :param abi_element: Function or Event ABI.
+    :param type: `ABIFunction` or `ABIEvent`
+    :return: Names for each input in the function or event ABI.
+    :rtype: `List[str]`
+    """
+    if (
+        "inputs" not in abi_element
+        or abi_element["type"] == "fallback"
+        or abi_element["type"] == "receive"
+    ):
+        return []
+    return [arg["name"] for arg in abi_element["inputs"]]
+
+
+def get_abi_input_types(abi_element: ABIElement) -> List[str]:
+    """
+    Return types for each input from the function or event ABI.
+
+    :param abi_element: Function or Event ABI.
+    :param type: `ABIFunction` or `ABIEvent`
+    :return: Types for each input in the function or event ABI.
+    :rtype: `List[str]`
+    """
+    if (
+        "inputs" not in abi_element
+        or abi_element["type"] == "fallback"
+        or abi_element["type"] == "receive"
+    ):
+        return []
```

Comment:
Agree with this. These functions were private in web3.py but exposing them seems reasonable to now raise properly. Handling those exceptions in web3 will be done later where encoding and decoding of arguments occurs. I'll update these to raise appropriately.

This would potentially break anything downstream that throws the `fallback` or `receive` function at it, but I think it's good to be explicit about why input types are not available.

For reference, this calls into the private method: https://github.com/ethereum/web3.py/blob/main/web3/contract/base_contract.py#L806